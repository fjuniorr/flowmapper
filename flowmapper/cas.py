from functools import cached_property

class CAS():
    """
    Class for CAS Registry Numbers® (CAS RN®) that accepts padded or non-padded strings
    """
    def __init__(self, cas: str):
        if not isinstance(cas, str):
            raise TypeError(f'cas should be a str, not {type(cas).__name__}')
        else:
            self.cas = cas.strip().lstrip('0')
            self.digits = tuple(int(d) for d in self.cas[0:-1].replace('-', ''))
    def __repr__(self):
        return self.cas
    def __eq__(self, other):
        if not self.cas and not other.cas:
                result = False
        else:
            result = self.cas == other.cas
        return result
    @cached_property
    def check_digit_expected(self):
        """
        Expected digit acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        result = sum([
            index * value 
            for index, value in enumerate(self.digits[::-1], start=1)
        ]) % 10
        return result
    @property
    def check_digit_actual(self):
        int(self.cas[-1])
    @property
    def valid(self):
        """
        True if check if CAS number is valid acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        return self.check_digit_actual == self.check_digit_expected
