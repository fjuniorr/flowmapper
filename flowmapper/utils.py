import json
from pathlib import Path
import hashlib
try:
    import tomllib
except ModuleNotFoundError:
    import tomli as tomllib

def read_field_mapping(filepath: Path):
    with open(filepath, 'rb') as file:
        data = tomllib.load(file)
    fields = data['fields']

    result = {
        'source': {},
        'target': {}
    }

    for key, values in fields.items():
        if len(values) == 1:
            result['source'][key] = values[0]
            result['target'][key] = values[0]
        else:
            result['source'][key] = values[0]
            result['target'][key] = values[1]
    
    return result


def generate_flow_id(flow: dict):
    flow_str = json.dumps(flow, sort_keys=True)
    result = hashlib.md5(flow_str.encode('utf-8')).hexdigest()
    return result


def read_flowlist(filepath: Path):
    with open(filepath, 'r') as fs:
        result = json.load(fs)
    return result
