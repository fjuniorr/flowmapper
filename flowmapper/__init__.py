"""flowmapper."""

from .flowmap import Flowmap

__all__ = (
    "__version__",
    "Flowmap"
    # Add functions and variables you want exposed in `flowmapper.` namespace here
)

__version__ = "0.0.0-9010"
