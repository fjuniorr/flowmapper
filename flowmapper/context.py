from dataclasses import asdict, dataclass

import flowmapper.jsonpath as jp


@dataclass
class Context:
    full: str = None
    primary: str = None
    secondary: str = None

    @classmethod
    def from_dict(cls, d, fields):
        fields = [fields] if isinstance(fields, str) else fields
        context_mapping_length = len(fields)
        primary_context = jp.extract(fields[0], d)
        if context_mapping_length == 1:
            result = Context(primary_context, primary_context)
        elif context_mapping_length == 2:
            secondary_context = jp.extract(fields[1], d)
            secondary_context = secondary_context if isinstance(secondary_context, str) else '/'.join(secondary_context)
            result = Context(
                full = f"{primary_context}/{secondary_context}",
                primary = primary_context,
                secondary = secondary_context
            )
        return result

    def to_dict(self):
        return asdict(self)
