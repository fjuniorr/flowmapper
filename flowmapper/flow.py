import flowmapper.jsonpath as jp

from .unit import Unit
from .cas import CAS
from .context import Context
from .utils import generate_flow_id, transform_flow
from .flowproperty import FlowProperty


class Flow:
    def __init__(
        self,
        original,
        fields=None,
        transformations=None,
    ):
        transformed = (
            transform_flow(original, transformations) if transformations else original
        )
        fields = fields if fields else {k:k for k in original}
        uuid_spec = fields.get("uuid")
        name_spec = fields.get("name")
        synonyms_spec = fields.get("synonyms")
        context_spec = fields.get("context")
        unit_spec = fields.get("unit")
        cas_spec = fields.get("cas")

        self.id = generate_flow_id(original)
        self.uuid = jp.extract(uuid_spec, transformed)
        self.uuid_raw_value = jp.extract(uuid_spec, original)
        self.uuid_raw_object = jp.extract(uuid_spec, original, "object")
        self.name = FlowProperty.from_dict(transformed, name_spec)
        self.name_raw_value = jp.extract(name_spec, original)
        self.name_raw_object = jp.extract(name_spec, original, "object")
        self.synonyms = FlowProperty.from_dict(transformed, synonyms_spec)
        self.synonyms_raw_value = jp.extract(synonyms_spec, original)
        self.synonyms_raw_object = jp.extract(synonyms_spec, original, "object")
        self.context = Context.from_dict(transformed, context_spec)
        self.context_raw_value = jp.extract(context_spec, original)
        self.context_raw_object = jp.extract(context_spec, original, "object")
        self.unit = Unit.from_dict(transformed, fields["unit"])
        self.unit_raw_value = jp.extract(unit_spec, original)
        self.unit_raw_object = jp.extract(unit_spec, original, "object")
        self.cas = CAS(jp.extract(cas_spec, transformed))
        self.cas_raw_value = jp.extract(cas_spec, original)
        self.cas_raw_object = jp.extract(cas_spec, original, "object")
        self.raw = original
        self.transformed = transformed

    def __repr__(self) -> str:
        return f"{self.name} (in {self.unit.value}) <{self.context.value}>"

    def __eq__(self, other):
        return self.id == other.id

    def __hash__(self):
        return hash(self.id)

    def __lt__(self, other):
        return self.name.raw_value < other.name.raw_value
