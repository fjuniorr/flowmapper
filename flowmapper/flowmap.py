from functools import cached_property
from .flow import Flow
from .match import match_rules, format_match_result
from .unit import Unit
from tqdm import tqdm
from typing import Callable, Optional
import pandas as pd
from pathlib import Path
import json
from collections import Counter

class Flowmap:
    """
    Crosswalk of flows from a source flow list to a target flow list.

    This class provides functionalities to map flows between different flow lists using a series of predefined match rules.

    Attributes
    ----------
    source_flows : list[Flow]
        The list of (unique) source flows to be mapped.
    source_flows_nomatch : list[Flow]
        The list of (unique) source flows that do not match any rule.
    target_flows : list[Flow]
        The list of target flows for mapping.
    target_flows_nomatch : list[Flow]
        The list of target flows that do not match any rule.

    """
    def __init__(
        self,
        source_flows: list[Flow],
        target_flows: list[Flow],
        rules: list[Callable[..., bool]] = None,
        nomatch_rules: list[Callable[..., bool]] = None,
        disable_progress: bool = False,
    ):
        """
        Initializes the Flowmap with source and target flows, along with optional matching rules.

        Duplicated flows are removed from both source and targets lists.

        Parameters
        ----------
        source_flows : list[Flow]
            The list of source flows to be mapped.
        target_flows : list[Flow]
            The list of target flows for mapping.
        rules : list[Callable[..., bool]], optional
            Custom rules for matching source flows to target flows. Default is the set of rules defined in `match_rules`.
        nomatch_rules : list[Callable[..., bool]], optional
            Rules to identify flows that should not be matched.
        disable_progress : bool, optional
            If True, progress bar display during the mapping process is disabled.

        """
        self.disable_progress = disable_progress
        self.rules = rules if rules else match_rules()
        if nomatch_rules:
            self.source_flows = []
            self.source_flows_nomatch = []
            
            for flow in source_flows:
                matched = False
                for rule in nomatch_rules:
                    if rule(flow):
                        self.source_flows_nomatch.append(flow)
                        matched = True
                        break
                if not matched:
                    self.source_flows.append(flow)
            self.source_flows = list(dict.fromkeys(self.source_flows))
            self.source_flows_nomatch = list(dict.fromkeys(self.source_flows_nomatch))

            self.target_flows = []
            self.target_flows_nomatch = []

            for flow in target_flows:
                matched = False
                for rule in nomatch_rules:
                    if rule(flow):
                        self.target_flows_nomatch.append(flow)
                        matched = True
                        break
                if not matched:
                    self.target_flows.append(flow)
            self.target_flows = list(dict.fromkeys(self.target_flows))
            self.target_flows_nomatch = list(dict.fromkeys(self.target_flows_nomatch))
        else:
            self.source_flows = list(dict.fromkeys(source_flows))
            self.source_flows_nomatch = []
            self.target_flows = list(dict.fromkeys(target_flows))
            self.target_flows_nomatch = []

    @cached_property
    def mappings(self):
        """
        Generates and returns a list of mappings from source flows to target flows based on the defined rules.

        Each mapping includes the source flow, target flow, conversion factor, the rule that determined the match, and additional information.

        A single match using the match rule with highest priority is returned for each source flow.

        Returns
        -------
        list[dict]
            A list of dictionaries containing the mapping details.

        """
        all_mappings = []
        for s in tqdm(self.source_flows, disable=self.disable_progress):
            for t in self.target_flows:
                for rule in self.rules:
                    is_match = rule(s, t)
                    if is_match:
                        all_mappings.append(
                            {'from': s,
                             'to': t,
                             'conversion_factor': s.conversion_factor if s.conversion_factor else s.unit.conversion_factor(t.unit),
                             'match_rule': rule.__name__,
                             'match_rule_priority': self.rules.index(rule),
                             'info': is_match}
                        )
                        break
        result = []
        seen_sources = set()
        sorted_mappings = sorted(all_mappings, key=lambda x: (x['from'], x['match_rule_priority']))
        for mapping in sorted_mappings:
            if mapping['from'] not in seen_sources:
                result.append(mapping)
                seen_sources.add(mapping['from'])
        
        return result

    @cached_property
    def _matched_source_flows_ids(self):
        return {map_entry['from'].id for map_entry in self.mappings}

    @cached_property
    def _matched_target_flows_ids(self):
        return {map_entry['to'].id for map_entry in self.mappings}

    @cached_property
    def matched_source(self):
        """
        Provides a list of source flows that have been successfully matched to target flows.

        Returns
        -------
        list[Flow]
            A list of matched source flow objects.

        """
        result = [
            flow
            for flow in self.source_flows 
            if flow.id in self._matched_source_flows_ids
        ]
        return result

    @cached_property
    def unmatched_source(self):
        """
        Provides a list of source flows that have not been matched to any target flows.

        Returns
        -------
        list[Flow]
            A list of unmatched source flow objects.

        """
        result = [
            flow 
            for flow in self.source_flows 
            if flow.id not in self._matched_source_flows_ids
        ]
        return result

    @cached_property
    def matched_source_statistics(self):
        """
        Calculates statistics for matched source flows, including the number of matches and the matching percentage for each context.

        Returns
        -------
        pandas.DataFrame
            A DataFrame containing matching statistics for source flows.

        """
        matched = Counter([flow.context.value for flow in self.matched_source])
        matched = pd.Series(matched).reset_index()
        matched.columns = ['context', 'matched']

        total = Counter([flow.context.value for flow in self.source_flows])
        total = pd.Series(total).reset_index()
        total.columns = ['context', 'total']

        df = pd.merge(matched, total, on='context', how='outer')
        df = df.fillna(0).astype({'matched': 'int', 'total': 'int'})

        df['percent'] = df.matched / df.total
        result = df.sort_values('percent')
        return result

    @cached_property
    def matched_target(self):
        """
        Provides a list of target flows that have been successfully matched to source flows.

        Returns
        -------
        list[Flow]
            A list of matched target flow objects.

        """
        result = [
            flow
            for flow in self.target_flows 
            if flow.id in self._matched_target_flows_ids
        ]
        return result

    @cached_property
    def unmatched_target(self):
        """
        Provides a list of target flows that have not been matched to any source flows.

        Returns
        -------
        list[Flow]
            A list of unmatched target flow objects.

        """
        result = [
            flow
            for flow in self.target_flows 
            if flow.id not in self._matched_target_flows_ids
        ]
        return result

    @cached_property
    def matched_target_statistics(self):
        """
        Calculates statistics for matched target flows, including the number of matches and the matching percentage for each context.

        Returns
        -------
        pandas.DataFrame
            A DataFrame containing matching statistics for target flows.

        """
        matched = Counter([flow.context.value for flow in self.matched_target])
        matched = pd.Series(matched).reset_index()
        matched.columns = ['context', 'matched']

        total = Counter([flow.context.value for flow in self.target_flows])
        total = pd.Series(total).reset_index()
        total.columns = ['context', 'total']

        df = pd.merge(matched, total, on='context', how='outer')
        df = df.fillna(0).astype({'matched': 'int', 'total': 'int'})

        df['percent'] = df.matched / df.total
        result = df.sort_values('percent')
        return result

    def statistics(self):
        """
        Prints out summary statistics for the flow mapping process.

        """
        source_msg = (
            f"{len(self.source_flows)} source flows ({len(self.source_flows_nomatch)} excluded)..."
            if self.source_flows_nomatch
            else f"{len(self.source_flows)} source flows..."
        )
        print(source_msg)
        target_msg = (
            f"{len(self.target_flows)} target flows ({len(self.target_flows_nomatch)} excluded)..."
            if self.target_flows_nomatch
            else f"{len(self.target_flows)} target flows..."
        )
        print(target_msg)
        print(
            f"{len(self.mappings)} mappings ({len(self.matched_source) / len(self.source_flows):.2%} of total)."
        )
        cardinalities = dict(Counter([x['cardinality'] for x in self._cardinalities]))
        print(f"Mappings cardinalities: {str(cardinalities)}")

    @cached_property
    def _cardinalities(self):
        """
        Calculates and returns the cardinalities of mappings between source and target flows.

        Returns
        -------
        list[dict]
            A sorted list of dictionaries, each indicating the cardinality relationship between a pair of source and target flows.

        """
        mappings = [(mapentry['from'].id, mapentry['to'].id) for mapentry in self.mappings]
        lhs_counts = Counter([pair[0] for pair in mappings])
        rhs_counts = Counter([pair[1] for pair in mappings])

        result = []

        for lhs, rhs in mappings:
            lhs_count = lhs_counts[lhs]
            rhs_count = rhs_counts[rhs]
            if lhs_count == 1 and rhs_count == 1:
                result.append({"from": lhs, "to": rhs, "cardinality": "1:1"})
            elif lhs_count == 1 and rhs_count > 1:
                result.append({"from": lhs, "to": rhs, "cardinality": "N:1"})
            elif lhs_count > 1 and rhs_count == 1:
                result.append({"from": lhs, "to": rhs, "cardinality": "1:N"})
            elif lhs_count > 1 and rhs_count > 1:
                result.append({"from": lhs, "to": rhs, "cardinality": "N:M"})

        return sorted(result, key = lambda x: x['from'])

    def to_randonneur(self, path: Optional[Path] = None):
        """
        Export mappings using randonneur data migration file format.

        Parameters
        ----------
        path : Path, optional
            If provided export the output file to disk.

        Returns
        -------
        list[dict]
            A list of dictionaries representing the formatted mapping results.

        """
        result = [
            format_match_result(map_entry['from'], 
                                map_entry['to'],
                                map_entry['conversion_factor'],
                                map_entry['info']) 
            for map_entry in self.mappings
        ]

        if not path:
            return result
        else:
            path = Path(path)
            path.parent.mkdir(parents=True, exist_ok=True)
            with open(path, 'w') as fs: 
                json.dump(result, fs, indent=2)
        

    def to_glad(self, path: Optional[Path] = None, ensure_id: bool = False):
        """
        Export mappings using GLAD flow mapping format, optionally ensuring each flow has an identifier.

        Formats the mapping results according to Global LCA Data Access (GLAD) network initiative flow mapping format.

        Parameters
        ----------
        path : Path, optional
            If provided export the output file to disk.
        ensure_id : bool, optional
            If True, ensures each flow has an identifier, default is False.

        Returns
        -------
        pandas.DataFrame
            A DataFrame containing the formatted mapping results in GLAD format.

        """
        data = []
        for map_entry in self.mappings:
            source_flow_id = map_entry['from'].uuid_raw_value if map_entry['from'].uuid_raw_value or not ensure_id else map_entry['from'].id
            row = {
                    'SourceFlowName': map_entry['from'].name_raw_value,
                    'SourceFlowUUID': source_flow_id,
                    'SourceFlowContext': '/'.join(map_entry['from'].context_raw_value),
                    'SourceUnit': map_entry['from'].unit_raw_value,
                    'MatchCondition': '',
                    'ConversionFactor': map_entry['conversion_factor'],
                    'TargetFlowName': map_entry['to'].name_raw_value,
                    'TargetFlowUUID': map_entry['to'].uuid_raw_value,
                    'TargetFlowContext': map_entry['to'].context.raw_value,
                    'TargetUnit': map_entry['to'].unit.raw_value,
                    'MemoMapper': map_entry['info'].get('comment')
                }
            data.append(row)

        result = pd.DataFrame(data)

        if not path:
            return result
        else:
            path = Path(path)
            path.parent.mkdir(parents=True, exist_ok=True)
            result.to_excel(path, index = False)
