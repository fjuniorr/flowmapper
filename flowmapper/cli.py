import typer
from typing_extensions import Annotated
from pathlib import Path
import logging
from typing import Optional
from typing_extensions import Annotated
import importlib.metadata
import logging
import json
from pathlib import Path
from enum import Enum
from randonneur import migrate_datasets
from .flow import Flow
from .utils import read_field_mapping, read_flowlist, read_migration_files
from .flowmap import Flowmap
try:
    import tomllib
except ModuleNotFoundError:
    import tomli as tomllib

logger = logging.getLogger(__name__)

app = typer.Typer()

def version_callback(value: bool):
    if value:
        print(f"flowmapper, version {importlib.metadata.version('flowmapper')}")
        raise typer.Exit()

class OutputFormat(str, Enum):
    all = 'all'
    glad = 'glad'
    randonneur = 'randonneur'

@app.callback()
def main(version: Annotated[
        Optional[bool],
        typer.Option("--version", callback=version_callback, is_eager=True),
    ] = None,):
    """
    Generate mappings between elementary flows lists
    """

@app.command()
def map(
    source: Annotated[Path, typer.Argument(help='Path to source flowlist')],
    target: Annotated[Path, typer.Argument(help='Path to target flowlist')],
    fields: Annotated[Path, typer.Option(help='Relationship between fields in source and target flowlists')],
    output_dir: Annotated[Path, typer.Option(help='Directory to save mapping and diagnostics files')] = Path('.'),
    format: Annotated[OutputFormat, typer.Option(help='Mapping file output format', case_sensitive=False)] = 'all',
    transformations: Annotated[list[Path], typer.Option("--transformations", "-t", help='Randonneur data migration file with changes to be applied to source flows before matching')] = None,
    unmatched_source: Annotated[bool, typer.Option(help='Write original source unmatched flows into separate file?')] = True,
    unmatched_target: Annotated[bool, typer.Option(help='Write original target unmatched flows into separate file?')] = True,
    matched_source: Annotated[bool, typer.Option(help='Write original source matched flows into separate file?')] = False,
    matched_target: Annotated[bool, typer.Option(help='Write original target matched flows into separate file?')] = False,
):
    """
    Generate mappings between elementary flows lists
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    field_mapping = read_field_mapping(fields)
    if transformations:
            transformations = read_migration_files(*transformations)
    source_flows = [Flow(flow, field_mapping['source'], transformations) for flow in read_flowlist(source)]
    target_flows = [Flow(flow, field_mapping['target']) for flow in read_flowlist(target)]

    flowmap = Flowmap(source_flows, target_flows)
    flowmap.statistics()

    stem = f'{source.stem}-{target.stem}'

    if matched_source:
        with open(output_dir / f'{stem}-matched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_source], fs, indent=True)

    if unmatched_source:
        with open(output_dir / f'{stem}-unmatched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_source], fs, indent=True)

    if matched_target:
        with open(output_dir / f'{stem}-matched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_target], fs, indent=True)

    if unmatched_target:
        with open(output_dir / f'{stem}-unmatched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_target], fs, indent=True)

    if format.value == 'randonneur':
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
    elif format.value == 'glad':
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
    else:
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
