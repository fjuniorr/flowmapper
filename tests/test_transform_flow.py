from flowmapper.utils import read_migration_files, read_flowlist
from flowmapper.flow import Flow
from flowmapper.flowmap import Flowmap

def test_transform_flow_with_conversion_factor(field_mapping, snapshot):
    
    source_flows_path = 'tests/data/sp.json'
    target_flows_path = 'tests/data/ei-3.7.json'
    
    transformations = read_migration_files('tests/data/transformations.json')
    
    source_flows = read_flowlist(source_flows_path)
    source_flows = [Flow(flow, field_mapping['source'], transformations) for flow in source_flows]
    target_flows = read_flowlist(target_flows_path)
    target_flows = [Flow(flow, field_mapping['target']) for flow in target_flows]

    flowmap = Flowmap(source_flows, target_flows)
    actual = flowmap.to_randonneur()
    assert actual == snapshot
