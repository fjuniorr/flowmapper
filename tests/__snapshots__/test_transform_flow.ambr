# serializer version: 1
# name: test_transform_flow_with_conversion_factor
  list([
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Air',
          '(unspecified)',
        ]),
        'name': '1,4-Butanediol',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Air',
          'high. pop.',
        ]),
        'name': '1,4-Butanediol',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.2142857142857142,
      'source': dict({
        'categories': list([
          'Emissions to air',
          'low. pop.',
        ]),
        'name': 'Ammonia, as N',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'name': 'Ammonia',
        'unit': 'kg',
        'uuid': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
      }),
    }),
    dict({
      'comment': 'Names with country code',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Emissions to air',
          'low. pop.',
        ]),
        'name': 'Ammonia, FR',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'location': 'FR',
        'name': 'Ammonia',
        'unit': 'kg',
        'uuid': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
      }),
    }),
  ])
# ---
